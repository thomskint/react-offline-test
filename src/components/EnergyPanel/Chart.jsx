import React from "react";
import PropTypes from "prop-types";
import ChartistGraph from "react-chartist";
import Typography from "@material-ui/core/Typography";
import { makeStyles } from "@material-ui/core/styles";
const useStyles = makeStyles(theme => ({
  graphStyle: {
    "& .ct-label": {
      fill: "black",
      fontWeight: "bold",
      fontSize: "1em",
      textTransform: "capitalize"
    }
  }
}));

const Chart = ({ generationmix }) => {
  const classes = useStyles();
  if (!generationmix)
    return <Typography variant="subtitle1">Error Creating Graph</Typography>;

  const data = {
    labels: generationmix.map(item => item.fuel),
    series: generationmix.map(item => item.perc)
  };

  const responsiveOptions = [
    [
      "screen and (min-width: 640px)",
      {
        chartPadding: 30,
        height: 500,
        labelOffset: 100,
        labelDirection: "explode",
        labelInterpolationFnc: function(value, idx) {
          return `${value} ${generationmix[idx].perc} %`;
        }
      }
    ],
    [
      "screen and (min-width: 1024px)",
      {
        labelOffset: 100,
        chartPadding: 20
      }
    ]
  ];

  const type = "Pie";
  return (
    <ChartistGraph
      className={classes.graphStyle}
      data={data}
      type={type}
      responsiveOptions={responsiveOptions}
    />
  );
};

PropTypes.Chart = {
  generationmix: PropTypes.array
};

export default Chart;
