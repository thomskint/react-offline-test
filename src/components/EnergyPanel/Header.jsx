import React from "react";
import PropTypes from "prop-types";
import CardHeader from "@material-ui/core/CardHeader";
const dateFormatter = dt => {
  try {
    return new Date(dt).toUTCString();
  } catch (e) {
    return "Incorrect Date";
  }
};
const Header = ({ to, from, classes }) => {
  return (
    <CardHeader
      className={classes.title}
      title={"Energy Generation Data UK"}
      subheader={`From: ${dateFormatter(from)} - To: ${dateFormatter(to)}`}
    />
  );
};

Header.propTypes = {
  to: PropTypes.string,
  from: PropTypes.string
};

export default Header;
