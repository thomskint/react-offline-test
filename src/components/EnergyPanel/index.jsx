import React, { useState } from "react";
import useDataFetching from "./../useDataFetching";
import Header from "./Header";
import Chart from "./Chart";
import { makeStyles } from "@material-ui/core/styles";
import Card from "@material-ui/core/Card";
import CardContent from "@material-ui/core/CardContent";
import Typography from "@material-ui/core/Typography";

const useStyles = makeStyles(theme => ({
  card: {
    minWidth: 800,
    minHeight: 650
  },
  title: {
    color: "#89BD23"
  }
}));

const EnergyPanel = () => {
  const { loading, results, error } = useDataFetching(
    "https://api.carbonintensity.org.uk/generation"
  );
  const classes = useStyles();

  if (loading || error) {
    return (
      <Card className={classes.card}>
        <CardContent>
          <Typography variant="subtitle1">
            {loading ? "Loading..." : error}
          </Typography>
        </CardContent>
      </Card>
    );
  }

  return (
    <Card className={classes.card}>
      {results.data && (
        <Header
          classes={classes}
          to={results.data.to}
          from={results.data.from}
        />
      )}
      <CardContent>
        {results.data && <Chart generationmix={results.data.generationmix} />}
      </CardContent>
    </Card>
  );
};

export default EnergyPanel;
