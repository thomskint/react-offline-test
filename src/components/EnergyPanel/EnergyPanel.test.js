import React from "react";
import { shallow, mount } from "enzyme";
import jsonData from "./../../../example_api_response.json";
import EnergyPanel from "./index";

import { act } from "react-dom/test-utils";

import useDataFetching from "./../useDataFetching";
describe("<EnergyPanel/> Component ", () => {
  beforeEach(() => {
    fetch.resetMocks();
  });

  it("renders correctly", () => {
    const wrapper = shallow(<EnergyPanel />);
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  it("when mounted the panel fetches chart data from server via the hook ", () => {
    const wrapper = mount(<EnergyPanel />);
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(
      "https://api.carbonintensity.org.uk/generation"
    );
    wrapper.unmount();
  });

  it("shows an error message if there is fetch error", () => {
    fetch.mockReject(new Error("nasty error message occured"));
    const wrapper = mount(<EnergyPanel />);

    expect(wrapper.html()).toEqual(
      '<div class="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-1 MuiPaper-rounded"><div class="MuiCardContent-root"><h6 class="MuiTypography-root MuiTypography-subtitle1">Loading...</h6></div></div>'
    );
    wrapper.unmount();
  });

  test("should render the loading message if its loading", () => {
    fetch.mockResponse(JSON.stringify(jsonData));
    const wrapper = mount(<EnergyPanel />);
    expect(wrapper.html()).toEqual(
      '<div class="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-1 MuiPaper-rounded"><div class="MuiCardContent-root"><h6 class="MuiTypography-root MuiTypography-subtitle1">Loading...</h6></div></div>'
    );
    wrapper.unmount();
  });
});
