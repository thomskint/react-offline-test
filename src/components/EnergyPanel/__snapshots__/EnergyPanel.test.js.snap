// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EnergyPanel/> Component  renders correctly 1`] = `ShallowWrapper {}`;
